### ISSUE TITLE
Error in ldap3 v2.0.9 when querying MS AD through LDAPs

### ISSUE THREAD

**Issue by kanayak123** [cannatag/ldap3#242]
I get this error in ldap3 2.0.9 when querying Microsoft AD. The same code was working till v1.4.0. Is it a bug or any extra config that I am missing. Couldn't find anything relevant in the [docs](http://ldap3.readthedocs.io/)

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/ldap3/core/connection.py", line 318, in __init__
    self.open(read_server_info=False)
  File "/usr/local/lib/python3.5/dist-packages/ldap3/strategy/sync.py", line 55, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.5/dist-packages/ldap3/strategy/base.py", line 139, in open
    raise exception_history[0][1](exception_history[0][2])
ldap3.core.exceptions.LDAPSocketOpenError: (LDAPSocketOpenError('socket ssl wrapping error: [SSL] PEM lib (_ssl.c:2825)',),)
```
```Server(host='10.xxx.xxx.xxx', port=636, use_ssl=True, allowed_referral_hosts=[('*', True)], tls=Tls(local_certificate_file='/etc/ssl/certs/example_local_cert.cer', validate=0), get_info='SCHEMA')```

`conn = ldap3.Connection(server, 'username', password = 'pass', auto_bind=True)`

I use Python 3.5

**Comment by cannatag** [comment]
This seems an error in the ssl library. Can you try to define the server object with ``get_info=NONE``? Also can you try to not verify the certificate in the Tls object, setting the ``validate`` parameter to ``ssl.CERT_NONE``?

**Comment by kanayak123** [comment]
Tried with both, did not work.

```
>>> server = ldap3.Server(host='10.xxx.xxx.xxx', port=636, use_ssl=True, allowed_referral_hosts=[('*', True)], tls=ldap3.Tls(local_certificate_file='/etc/ssl/certs/example_local_cert.cer', validate=ssl.CERT_NONE), get_info=None)
>>> conn = ldap3.Connection(server, 'username', password = 'pass', auto_bind=True)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/ldap3/core/connection.py", line 321, in __init__
    self.open(read_server_info=False)
  File "/usr/local/lib/python3.5/dist-packages/ldap3/strategy/restartable.py", line 56, in open
    SyncStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.5/dist-packages/ldap3/strategy/sync.py", line 55, in open
    BaseStrategy.open(self, reset_usage, read_server_info)
  File "/usr/local/lib/python3.5/dist-packages/ldap3/strategy/base.py", line 139, in open
    raise exception_history[0][1](exception_history[0][2])
ldap3.core.exceptions.LDAPMaximumRetriesError: LDAPMaximumRetriesError: LDAPMaximumRetriesError: restartable connection strategy failed while opening socket/Exception history:/    0 2016-11-07 16:57:29.168776: <class 'ldap3.core.exceptions.LDAPSocketOpenError'> - ('socket ssl wrapping error: [SSL] PEM lib (_ssl.c:2825)',)/Maximum number of retries reached: 30
>>>
```
Same code worked fine with ldap3 v1.4

**Comment by cannatag** [comment]
can you try with the SYNC strategy (client_strategy=SYNC in Connection(...)) and the log_level set to EXTENDED? Then please send me the log file:

```
import logging
logging.basicConfig(filename='client_application.log', level=logging.DEBUG)
from ldap3.utils.log import set_library_log_activation_level
set_library_log_activation_level(logging.DEBUG)
set_library_log_detail_level(EXTENDED)
```

**Comment by kanayak123** [comment]
Attached logs for both ldap3 v2 and ldap3 v1.4

[ldap3.zip](https://github.com/cannatag/ldap3/files/577874/ldap3.zip)

Please let me know if I can be of anymore help..

**Comment by kanayak123** [comment]
Attached both v2.1.0 and v1.4.0 logs, for your comparison.
[ldap3.zip](https://github.com/cannatag/ldap3/files/577928/ldap3.zip)

Let me know if I can be of anymore help.

**Comment by cannatag** [comment]
thanks, Are you using the same python interpreter? Which version?

**Comment by kanayak123** [comment]
The logs are from same client, just reinstalled ldap3 for different versions. Python 3.5.2 is installed on it.

**Comment by cannatag** [comment]
Are you using a certificate file? This seems related to a problem with the certificates chain. Version 2 has the abilitity to use a certificate file that contains the cert and the key. That's the difference with version 1.4.

Look at issue #203 and to PR #204. You can try to revert PR in the ldap3/core/tls.py changing:
```
if self.certificate_file:
```
with
```
if self.private_key_file:
```

at line 176.

**Comment by kanayak123** [comment]
Yes, my Active Directory Domain Controller requires me to use a certificate file to be able to use LDAPs and I believe its a default behavior in MS AD. Also additionally, MS Active Directory does not allow any modifications to any attributes in AD, unless the commands are sent over LDAPs.

About the above issue, your suggestions worked perfectly. I can now query AD successfully ldap3 v2.1.0 with, after changing line 176 in `ldap3/core/tls.py` to  `if self.private_key_file: `

Are you going to change this in the next release?

**Comment by cannatag** [comment]
Hi, I think you're mixing up certificates: The server certificate is not needed when establishing the connection. The certificate in the Tls() object is the "user" certificate, and you need your private key too.

If your server uses a certificate issued internally ijn your company you must pass the CA certs to the Tls object if you want that ldap3 validates it or use the validate=ssl.CERT_NONE without passing any certificate to the Tls object.

I don't think I will revert the "if" condition in the tls.py file because it's correct. The certificate file could contain the private key, so to check the existence of a "user" certificate that is the right "if" condition.

**Comment by kanayak123** [comment]
Agree that my initial certificate parameters were wrong. But I still have issues even with correct certificates, Here is the description.

Scene1: #   No certificate defined `ldap3.Tls(validate=ssl.CERT_NONE)`
Query to the LDAPs is successful. Modify of any attributes with correct parameters failed with error unwillingToPerform. The same code works fine with ldap3 v1.4.

Scene2: Correct Client Certificate, Private Key and CA Certificate provided `ldap3.Tls(local_private_key_file = '/etc/ssl/private/example_local_key.pem', local_certificate_file = '/etc/ssl/certs/example_local_cert.pem', validate = ssl.CERT_REQUIRED, ca_certs_file = '/etc/ssl/certs/example_local_CA.cer')`
Couldn't query or modify

Scene3: Correct Client Certificate and Private key supplied with ssl.CERT_NONE. `ldap3.Tls(local_private_key_file = '/etc/ssl/private/example_local_key.pem', local_certificate_file = '/etc/ssl/certs/example_local_cert.pem', validate = ssl.CERT_NONE)`
Couldn't query or modify

The logs from all 3 scenarios are attached for your reference.
[ldap3.zip](https://github.com/cannatag/ldap3/files/588952/ldap3.zip)

**Comment by cannatag** [comment]
Scene1: "UnwillingToPerform" means that the server has all information to execute the requested operation, but there is some requirement that is not fulfilled. It could happen because your password doesn't follow the password policy or the connection is not secure.

Scene2 and Scene3: in both cases you've a problem with the certificate client configuration.

**Comment by kanayak123** [comment]
Scene1: You are right. Some mandatory requirement is not fulfilled when modifying the attribute, hence AD is rejecting the request. But, it is not the value that I am passing or in the server definition, because the same code & value woks with ldap v1.4. Please see the logs. It is from the same client. I have made 2 calls to 2 different Domain Controllers. Each call is from ldap3 v1.4 and v2.1. The **modify** calls with v1.4 succeeds but 2.1 fails.

Scene2 and Scene3: I will validate the certificates with some windows tools and confirm. These were generated and exported by referring to the MS Docs.

[ldap3.zip](https://github.com/cannatag/ldap3/files/589553/ldap3.zip)

I have anyway copied the entire code, if it helps.
```
import logging, ldap3, ssl

from ldap3.utils.log import set_library_log_activation_level, set_library_log_detail_level, EXTENDED
logging.basicConfig(filename='//var//log//client_application.log', level=logging.DEBUG)
set_library_log_activation_level(logging.DEBUG)
set_library_log_detail_level(EXTENDED)
username = 'search_user'
def search():
   server = ldap3.Server(host='10.xxx.xxx.xxx', port=636, use_ssl=True, allowed_referral_hosts=[('*', True)], tls=ldap3.Tls(validate=ssl.CERT_NONE), get_info=ldap3.NONE)
   conn = ldap3.Connection(server, 'username', password = 'pass', auto_bind=True, client_strategy=ldap3.SYNC)
   conn.search(search_base = 'DC=exmaple,DC=local',
         search_filter = '(sAMAccountName=%s)' %username,
         search_scope = ldap3.SUBTREE,
         attributes = ['cn', 'mail', 'employeeID', 'homePhone', 'pager', 'mobile'])
   print(conn.entries)
   print(conn.entries[0].mail.value)
   conn.modify('CN=Serach User,OU=users,OU=IT,DC=exmaple,DC=local',
         {'unicodePwd': [(ldap3.MODIFY_REPLACE, ['"\x00E\x00%\x003\x004\x003\x002\x00$\x002\x005\x004\x00j\x00h\x00g\x00h\x00"\x00'])]})
   print(conn.result)

search()
```

**Comment by dretrex** [comment]
I would just like to add that I have run into the exact same issue as what is described by Scene1.  I get unwilling to perform when using v2.1 and it works with v1.4 with no code change aside from changing the version of the package.

**Comment by cannatag** [comment]
I will set up a windows machine in my lab and will check 2.1 against 1.4.

**Comment by cannatag** [comment]
The correct signature for the extend.microsoft.modify_password is (dn, OLD_password, NEW_password). In case you're the Administrator you can set OLD_password to None, but it must be passed.

**Comment by kanayak123** [comment]
Tried that as well, but did not work. Observed 2 things.

When we use ldap3 v2.1 the encoded value is different than 1.4. Not sure if that is causing the problem. This encoded passoword does not work with 1.4 either.
`DEBUG:ldap3:PROTOCOL:MODIFY request <{'changes': [{'operation': 2, 'attribute': {'type': 'unicodePwd', 'value': ['"\\00J\\00H\\00g\\00s\\00j\\00h\\00g\\006\\007\\005\\007\\006\\005\\003\\005\\006\\004\\00%\\00$\\00%\\00"\\00']}}], 'entry': 'CN=User Name,OU=users,OU=IT,DC=example,DC=local'}> sent via `

Secondly, I even tried with the correct encoded value that works with 1.4. 
`DEBUG:ldap3:PROTOCOL:MODIFY request <{'changes': [{'attribute': {'type': 'unicodePwd', 'value': ['"\x00J\x00H\x00g\x00s\x00j\x00h\x00g\x006\x007\x005\x007\x006\x005\x003\x005\x006\x004\x00%\x00$\x00%\x00"\x00']}, 'operation': 2}], 'entry': 'CN=User Name,OU=users,OU=IT,DC=example,DC=local'}> sent`

But still I get the same error with ldap3 v2.1. `{'result': 53, 'type': 'modifyResponse', 'message': '0000001F: SvcErr: DSID-031A12D2, problem 5003 (WILL_NOT_PERFORM), data 0\n\x00', 'dn': '', 'referrals': None, 'description': 'unwillingToPerform'}`

I have spent around 2 hrs and tried all possible combination of passing the correct values. All combinations with ldap3 v2.1 returns error, and the same code works perfectly with v1.4

**Comment by cannatag** [comment]
Thanks Amith, this is very valuable to me. I've set up a AD lab and hope to fix it soon.

**Comment by cannatag** [comment]
check code in dev. It should work now.

**Comment by kanayak123** [comment]
Yes, working perfectly with signature `extend.microsoft.modify_password(dn, new_password)` and `extend.microsoft.modify_password(dn, old_password=None, new_password)`
Thank you...

**Comment by cannatag** [comment]
good. will be released in 2.1.1
